### PR TITLE
Delegating collection factories do not rebuild collections to be specific type the factory delegates to

### PR DESCRIPTION
**Pull request by joshlemer**
A lot of collections factories are set up as delegates which outsource their factory logic to an other factory. For example, `immutable.Seq` delegates to `immutable.List` and `collection.IndexedSeq` delegates to `immutable.Vector`. However, those more specific builders, when calling `.from(IterableOnce[E])`, will build a new collection of their specific type, if the value is not already of that type. This results in the following behaviour (among others):

```scala
scala> Seq.from(Vector(1,2,3))
res6: Seq[Int] = List(1, 2, 3)

scala> IndexedSeq.from(1 to 5)
res7: IndexedSeq[Int] = Vector(1, 2, 3, 4, 5)

scala> collection.immutable.LinearSeq.from(collection.immutable.Queue(1))
res17: scala.collection.immutable.LinearSeq[Int] = List(1)

```
This PR changes that so that we get instead:

```scala
scala> Seq.from(Vector(1,2,3))
res6: Seq[Int] = Vector(1, 2, 3)

scala> IndexedSeq.from(1 to 5)
res7: IndexedSeq[Int] = Range 1 to 5

scala> collection.immutable.LinearSeq.from(collection.immutable.Queue(1))
res17: scala.collection.immutable.LinearSeq[Int] = Queue(1)

```

### CODE REVIEW

**Review by NthPortal — APPROVED** [review]
LGTM. @julienrf?

@joshlemer it's probably worth having a couple tests to confirm (e.g. `Seq.from(1 to 10).isInstanceOf[Range]`)

**Comment by joshlemer** [comment]
@NthPortal sounds good :+1:

**Review by julienrf — APPROVED** [review]

**Comment by joshlemer** [comment]
@NthPortal @julienrf tests are added.

I wonder if we should change behavior `Delegate#apply` method?  Right now `Seq(1,2,3)` returns `List(1,2,3)`. After this PR, `Seq(1,2,3)` will return `ArraySeq(1,2,3)` because 

```scala
apply(as: A*) = from(a)
```

the varargs are already an `immutable.Seq` so they are not transformed to a `List` but instead the vararg is returned

Likewise, 
```scala
IndexedSeq(1,2,3)
// ArraySeq(1,2,3)
```
but before, this would return `Vector(1,2,3)`

Options:
1) also override `SeqFactory.Delegate#apply(as: A*) = delegate.apply(as: _*)` to instead point to `List.apply` and `Vector.apply` not `this.from`. So this will fix both Vector and List cases

2) Users should specify they want a Vector if they want a Vector, and a List if they want a List. That might be kinda surprising, and deserves release notes label I think.

**Comment by julienrf** [comment]
> Right now `Seq(1,2,3)` returns `List(1,2,3)`. After this PR, `Seq(1,2,3)` will return `ArraySeq(1,2,3)`

Oh, this is a problem. `Seq(1, 2, 3)` should instantiate the “default” `Seq` implementation, which is `List`, not `ArraySeq`.

I think the solution that consists of overriding `apply` is correct.

**Comment by joshlemer** [comment]
Ok that behavior is fixed now, and I also forward all the other `apply` methods of delegating factories (MapFactories, SortedIterableFactories, ..) to the delegate for consistency sake.

**Review by julienrf — APPROVED** [review]

File: src/library/scala/collection/immutable/SortedMap.scala [hunk 1]
```diff
@@ -86,6 +86,11 @@ trait SortedMapOps[K, +V, +CC[X, +Y] <: Map[X, Y] with SortedMapOps[X, Y, CC, _]
 @SerialVersionUID(3L)
 object SortedMap extends SortedMapFactory.Delegate[SortedMap](TreeMap) {
 
+  override def from[K: Ordering, V](it: IterableOnce[(K, V)]): SortedMap[K, V] = it match {
+    case sm: SortedMap[K, V] if implicitly[Ordering[K]] == sm.ordering => sm
```
Comment (NthPortal) [inline src/library/scala/collection/immutable/SortedMap.scala]: good catch here

File: src/library/scala/collection/SortedMap.scala [hunk 2]
```diff
@@ -223,4 +223,4 @@ object SortedMapOps {
 }
 
 @SerialVersionUID(3L)
-object SortedMap extends SortedMapFactory.Delegate[SortedMap](TreeMap)
+object SortedMap extends SortedMapFactory.Delegate[SortedMap](immutable.SortedMap)
```
Comment (NthPortal) [inline src/library/scala/collection/SortedMap.scala:223]: what's the reason for this change?

Comment (joshlemer) [reply src/library/scala/collection/SortedMap.scala:223]: We can return the argument collection from `from` if and only if the argument is an immutable sorted map. That means that the 'from' logic is identical to the immutable sorted map object, so we should either copy paste that 'from' implementation here, or just delegate to that one

File: test/junit/scala/collection/FactoriesTest.scala [hunk 3]
```diff
@@ -105,6 +110,177 @@ class FactoriesTest {
...
+      val x, y, z = Value
+    }
+
+    val enumValues = enum.values
+
+    val ord0 = enumValues.ordering
+    val ord1 = implicitly[Ordering[enum.Value]]
+
+    sortedFactoryFromIterableOnceReturnsSameReference(SortedSet, im.SortedSet)(enumValues)
+    */
```
Comment (NthPortal) [inline test/junit/scala/collection/FactoriesTest.scala]: what's up with this commented out block?

Comment (joshlemer) [reply test/junit/scala/collection/FactoriesTest.scala]: Currently, the ordering that is implicitly available on enumeration.Value, is a low priority Ordering that uses implicit conversions from `Comparable[T]`. However, the ordering used by the SortedSet `enum.values` uses a different ordering (`Enumeration#ValueOrdering`). So then, when you call `SortedSet.from(enum.values)` it picks up a different ordering than the one that that sortedSet uses. So it rebuilds the SortedSet with the low-priority ordering.

Comment (joshlemer) [reply test/junit/scala/collection/FactoriesTest.scala]: So, that will fail the test because the two orderings are new referentially equal (nor `==` equal)

Comment (joshlemer) [reply test/junit/scala/collection/FactoriesTest.scala]: @NthPortal see https://github.com/scala/scala/pull/7591

**Comment by joshlemer** [comment]
One of the tests are failing because in the scala.collection.generic, the implicitly available `IsSeq[String]` is
```scala
  implicit val stringIsSeq: IsSeq[String] { type A = Char; type C = String } =
    new IsSeq[String] {
      type A = Char
      type C = String
      def apply(s: String): SeqOps[Char, immutable.IndexedSeq, String] =
        new SeqOps[Char, immutable.ArraySeq, String] {
          def length: Int = s.length
          def apply(i: Int): Char = s.charAt(i)
          def toIterable: Iterable[Char] = new immutable.WrappedString(s)
          protected[this] def coll: String = s
          protected[this] def fromSpecific(coll: IterableOnce[Char]): String = coll.iterator.mkString
          def iterableFactory: IterableFactory[immutable.ArraySeq] = immutable.ArraySeq.untagged
          protected[this] def newSpecificBuilder: mutable.Builder[Char, String] = new StringBuilder
          def iterator: Iterator[Char] = s.iterator
        }
    }
```

so it returns the ArraySeq iterableFactory, not `IndexedSeq` or `Vector`. But the test expects `Vector`. Not sure what's right here.

**Comment by julienrf** [comment]
My understanding is that in that specific test, the end result is determined by the `BuildFrom` instance, which should be resolved to `BuildFrom.fallbackStringCanBuildFrom` and uses `immutable.IndexedSeq.from` to build its result. In your case, the `IsSeq` instance has built an `immutable.ArraySeq`, which is passed to the `immutable.IndexedSeq.from` method. With the changes of this PR, that method does not build a new collection if it is already an `immutable.IndexedSeq`, that’s why we get an `immutable.ArraySeq`.

I think it is fine to change the expectation of that test (so, you should update the `enrich-gentraversable.check` file to expect an `ArraySeq` instead of a `Vector`).

File: src/library/scala/collection/SortedMap.scala (hunk 2 above)
Comment (NthPortal) [reply src/library/scala/collection/SortedMap.scala:223]: the other ones all seem to delegate to the concrete collection

**Comment by Ichoran** [comment]
@joshlemer - Looks like there's a serialization stability test that's failing.  Not sure why this would have changed `mutable.TreeMap`'s serialization, though.  Any ideas?

File: src/library/scala/collection/SortedMap.scala (hunk 2 above)
Comment (joshlemer) [reply src/library/scala/collection/SortedMap.scala:223]: `collection.{Seq, LinearSeq, IndexedSeq}` also delegate to `immutable.{Seq, LinearSeq, IndexedSeq}`, though when you wrote this comment, `collection.LinearSeq` did delegate directly to `immutable.List`.

File: test/junit/scala/collection/FactoriesTest.scala (hunk 3 above)
Comment (joshlemer) [reply test/junit/scala/collection/FactoriesTest.scala]: No longer commented out, since #7591 is merged.

**Comment by joshlemer** [comment]
I honestly don't get how this change could hve affected the serialization of 

```scala
mutable.TreeMap[Int, Int]()
```

**Comment by javax-swing** [comment]
Looks like the current serialization closes over a lambda which has now changed. (or at least that's my guess). If you base 64 decoder the output from the test results you get the below.

old
```
�sr�2scala.collection.generic.DefaultSerializationProxy��������L�factoryt�Lscala/collection/Factory;xpsr�+scala.collection.SortedMapFactory$ToFactory��������L�evidence$41t�Lscala/math/Ordering;L�factoryt�#Lscala/collection/SortedMapFactory;xpsr�&scala.runtime.ModuleSerializationProxy��������L�moduleClasst�Ljava/lang/Class;xpvr�scala.math.Ordering$Int$?zPjj��xpsq�~�vr�!scala.collection.mutable.TreeMap$���������xpw����x
```

new 
```
�sr�2scala.collection.generic.DefaultSerializationProxy��������L�factoryt�Lscala/collection/Factory;xpsr�+scala.collection.SortedMapFactory$ToFactory��������L�evidence$42t�Lscala/math/Ordering;L�factoryt�#Lscala/collection/SortedMapFactory;xpsr�&scala.runtime.ModuleSerializationProxy��������L�moduleClasst�Ljava/lang/Class;xpvr�scala.math.Ordering$Int$?zPjj��xpsq�~�vr�!scala.collection.mutable.TreeMap$���������xpw����x
```

based on the above, i think the implicit evidence in the apply method is the culprit.
```
@SerialVersionUID(3L) class Delegate[CC[_, _]](delegate: SortedMapFactory[CC]) extends SortedMapFactory[CC] { override def apply[K: Ordering, V](elems: (K, V)*): CC[K, V] = delegate.apply(elems: _*)
```

`evidence$42t` rather than `evidence$41t`

**Comment by joshlemer** [comment]
Ah ok, probably safe to regenerate the serialization tests then eh?

**Comment by javax-swing** [comment]
From what i can tell. Its because implicit evidence seems to be numerically assigned a field name (based on all the evidence in the same file possibl). This field name has changed due to other implicit evidences. this suprised me.

**Comment by joshlemer** [comment]
Maybe to ensure this doesn't happen again, we should avoid context bound syntax?

**Comment by javax-swing** [comment]
Perhaps there should be a warning that using the syntax `class[A : TypeClass]` is not recommended for serializable classes

**Comment by Ichoran** [comment]
@joshlemer - Is this ready to go?  (Seem to me like yes?)

**Comment by joshlemer** [comment]
@Ichoran ready freddy
